Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,370 @@
+extenal command line programs like `clang` or `bash`. Typical projects contain
+many rules, targets, and BUILD files.
+
+```
+BUILD file -> target -> rule -> action -> execution
+```
+
+Bazel and open source rules should provide most functionality to build an iOS
+application. Generally, defining custom rules isn't required but can improve
+and consolidate functionality. To learn more about how to create custom rules,
+the document, [Bazel extensions](https://docs.bazel.build/versions/master/skylark/concepts.html)
+contains a comprehensive overview.
+	
+### Command line usage
+
+In Xcode, command line builds are achieved through the command line interface
+`xcodebuild`.  to build the scheme `ios-app` from `MyProject.xcworkspace`,
+`xcodebuild`, would be invoked as:
+
+```
+xcodebuild -workspace MyProject.xcworkspace -scheme ios-app
+```
+
+Bazel exposes a command line which can also do builds. In Bazel, _every_ target
+is in the global WORKSPACE and there is no notion of schemes.
+```
+bazel build ios-app
+```
+
+The Bazel command line has hundreds of options, which can be found in the
+[Bazel
+documentation](https://docs.bazel.build/versions/master/command-line-reference.html).
+For iOS developers, a set of useful flags is available at
+[bazel-ios-users](https://github.com/ios-bazel-users/ios-bazel-users/blob/master/UsefulFlags.md).
+
+In addition to the default bazel options, it's common to create custom
+configuration settings to customize builds, so the possible Bazel command line
+permutations are endless.
+
+For example, the following build file has conditional `copts` on `:app_store`
+
+```
+objc_library(
+   name = "some",
+   srcs = ["some.m"],
+   copts = select({
+      # For app_store we need to build `some.m` with a special copt
+      ":app_store": ["-DAPPSTORE=1"],
+      "//conditions:default": [],
+    })
+)
+config_setting(
+    name = "app_store",
+    values = { "app_store" : "true" }
+)
+```
+
+This is passed in as a define to Bazel
+```
+bazel build ios-app --define app_store=true
+```
+
+For more information about `config_setting`, please see the [Bazel
+documentation](https://docs.bazel.build/versions/master/be/common-definitions.html#configurable-attributes).
+
+### Configuration of libraries and flags via macros
+
+Most users of Bazel implement a higher level system of macros to encapsulate
+defaults of building librarys and simplify configuration management. 
+
+Bazel provides the pythonic programming language Starlark to implement such
+build system logic. Like rules and aspects, macros are defined in `.bzl` files.
+A macro is a convenient way to call a rule, and not recognized by Bazel in the
+same way a rule is.
+
+_Note: The main distinction between a `.bzl` and a `BUILD` file is `BUILD` files are
+used to create targets by calling macros and rules. `.bzl` files define the
+implementation._
+
+To create a wrapper for `objc_library`, create the file `objc_library.bzl`. The
```

Comment:
maybe it would be better to have the example macro not match the native `objc_library` rule name exactly? That seems like it might be more confusing.

You could have `pins_objc_library` for example and likely get the same points across

Reply:
I think this is an opinionated implementation choice. Personally, I think it's nice and also prevents people from re-using `objc_library` when they should be using the one that prevents overriding `copts`. I liked up the py rules in the original discussion as well.